### PR TITLE
Add contents entries for domain objects

### LINKED ISSUE #6316: Create a ToC entry for every function, method, class, etc
It would be useful to have an option that causes Sphinx to automatically create a TOC entry for every function, class, and method.  (In the absence of this, tables of contents are of limited value).

### LINKED ISSUE #10804: automodule places members under docstring headers
### Describe the bug

Whenever you use

```
.. automodule:: mod
   :members:
```

Sphinx inserts the module docstring, then inserts the members under that docstring. If the docstring contains headers, the functions are all placed under the bottommost header. 

This is hard to tell in most themes, because it isn't actually evident what lives under a given header. However, you can tell if you inspect the webpage.

This came up as I was working on creating an extension to add autodoc functions to the toctree (see https://gist.github.com/agoose77/e8f0f8f7d7133e73483ca5c2dd7b907f and https://github.com/sphinx-doc/sphinx/issues/6316). With this behavior, it places the functions under the module headers in the toctree, which is not what I want.

### How to Reproduce

I have created a small reproducer project here https://github.com/asmeurer/sphinx-automodule-test. There is a build here https://www.asmeurer.com/sphinx-automodule-test/

You can see if you inspect the page that `mod.function` is under `subheader`, and `mod.submod.function2` is not. 

In practice, this comes up in SymPy in several places, for example on [this doc page](https://docs.sympy.org/latest/modules/physics/wigner.html). With the current behavior and the extension I am working on, all the functions in that module end up under the "authors" header in the module docstring, whereas they ought to be at the top-level.

### Expected behavior

I dug into the code and it looks like the problem is that `.. automodule` with `:members:` simply generates RST like

```rst
.. module:: mod

<module docstring here ...>

Header
======

.. autofunction:: mod.function
```

Which tricks RST into thinking that the module headers are part of the top-level document.

It would be better if this instead put the module docstring as content of the `module` directive, like

```rst
.. module:: mod
   
   <module docstring here ...>
   
   Header
   ======

.. autofunction:: mod.function
```

However, `py:module` is different from every other directive in that it does not allow content. Is there a reason for this? If py:module worked like every other directive and allowed the docstring to be included as content, then something along the lines of

```diff
diff --git a/sphinx/ext/autodoc/__init__.py b/sphinx/ext/autodoc/__init__.py
index 931c14049..5037d340e 100644
--- a/sphinx/ext/autodoc/__init__.py
+++ b/sphinx/ext/autodoc/__init__.py
@@ -956,6 +956,12 @@ class ModuleDocumenter(Documenter):
         merge_members_option(self.options)
         self.__all__: Optional[Sequence[str]] = None

+    def add_content(self, more_content: Optional[StringList]) -> None:
+        old_indent = self.indent
+        self.indent += '   '
+        super().add_content(more_content)
+        self.indent = old_indent
+
     @classmethod
     def can_document_member(cls, member: Any, membername: str, isattr: bool, parent: Any
                             ) -> bool:
```

would fix this (not claiming that is the cleanest fix on the autodoc side, but I believe it would get the job done).

### Your project

https://github.com/asmeurer/sphinx-automodule-test

### Screenshots

_No response_

### OS

Mac

### Python version

3.9

### Sphinx version

master branch

### Sphinx extensions

sphinx.ext.autodoc

### Extra tools

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by AA-Turner**
Closes #6316, closes #10804

Per a recent request on the CPython discord, this adds entries in the table of contents for domain objects (e.g. `py:function`, `js:module`, etc).

* [Example page](https://sphinx--10807.org.readthedocs.build/en/10807/extdev/nodes.html)
* [Second example](https://sphinx--10807.org.readthedocs.build/en/10807/extdev/appapi.html)
* [Third example](https://sphinx--10807.org.readthedocs.build/en/10807/extdev/i18n.html)

It does not support objects from the C, C++, and RST domains, as well as the `option`, `cmdoption`, and `envvar` objects.

A [draft implementation](https://gist.github.com/agoose77/e8f0f8f7d7133e73483ca5c2dd7b907f) by @agoose77 and @asmeurer was useful background, thank you both for this.

### Feature or Bugfix
- Feature

### Relates
- https://github.com/python/cpython/issues/83545

Ping for feedback and review, from people who have commented on the linked issues/CPython Discord thread: @Phillip-M-Feldman @slateny @pradyunsg @CAM-Gerlach @tony @agoose77 @asmeurer

A

### CODE REVIEW

**Comment by asmeurer** [comment]
Great, this saves me from having to turn that gist into a real Sphinx extension. Does this also include support for glossary entries?

I will test this out tomorrow.

**Comment by jakobandersen** [comment]
Very nice. Can you outline the requirements for a domain to hook into / support this?

**Comment by n-peugnet** [comment]
Could this change also allow the standard domain to provide objects, like the `genindex` document, the same way https://github.com/sphinx-doc/sphinx/pull/10673 is trying to do it ?

**Comment by agoose77** [comment]
I've not tried this out yet, but thanks for tackling it. I've just had no time of late, so I really appreciate someone pushing it over the line!

**Comment by AA-Turner** [comment]
> Does this also include support for glossary entries?

No, should it? Such a thing could be a follow-up PR.

A

**Comment by AA-Turner** [comment]
> Very nice. Can you outline the requirements for a domain to hook into / support this?

Define the `fullname` attribute on the first child of an `addnodes.desc` object, such that:

```python
(isinstance(node, addnodes.desc) and bool(node[0]['fullname'])) is True
```

The python domain handles the definition in `PyObject.handle_signature()`, line 510.

A

**Comment by asmeurer** [comment]
I included glossary entries in the gist. It would be nice if they were also added to the to, although it's fine if that's a separate PR.

**Comment by asmeurer** [comment]
As I noted on the issue, I think we should keep the literal styling. This makes it easier to distinguish between functions and normal headers. Also, Sphinx always renders links to functions using `literal` styling, so this would be consistent with that.

**Comment by asmeurer** [comment]
As for showing the class name, I'm ambivalent. I can understand if it seems like too much, but it also makes it a little easier to see what's going on for classes with a lot of methods. Here's an example of part of the sidebar in the SymPy docs (built with my gist which includes the class name)
Screen Shot 2022-09-08 at 12 53 51 PM
Of course, if we can't decide on it, we can make it customizable. 

I agree we shouldn't include the fully qualified name (although in general I wish Sphinx had better ways of letting you omit the fully qualified part of a function name in the docs).

File: sphinx/domains/python.py [hunk 1]
```diff
@@ -984,7 +985,14 @@ def run(self) -> List[Node]:
         modname = self.arguments[0].strip()
         noindex = 'noindex' in self.options
         self.env.ref_context['py:module'] = modname
-        ret: List[Node] = []
+
+        content_node: Element = nodes.section()
+        with switch_source_input(self.state, self.content):
```
Comment (asmeurer) [inline sphinx/domains/python.py:1020]: Glad to see this fixed here too. I didn't look into it too deeply, but should `PyModule` just use the same mechanisms as the other Py* classes (subclassing from `PyObject`)?

Also I suppose the docs should be updated for this.

**Comment by asmeurer** [comment]
I wonder if people would want this whole thing to be disableable with a configuration option. I guess if no one asks for it it's probably better to not add it, but it does seem like it could be something someone might not want.

**Comment by asmeurer** [comment]
Can you cross-link the CPython Discord thread? I couldn't find any reference to it anywhere.

**Comment by asmeurer** [comment]
Oh is that *Discord*, not *Discourse*? I see why it wasn't linked if that's the case.

**Comment by AA-Turner** [comment]
I've updated the implementation:

- A new configuration variable `toc_object_entries_parents` (name suggestions welcome) controls how ToC entries should be displayed, with parents hidden (the default), the immediate parent shown, or all parents shown
- Formatting of names is delegated to domains, as not all languages show inheritance the same (e.g. `.` vs `->`)
- To support displaying table of contents entries, domains must put the inheritance path of each object in the  `self._toc_parents` tuple (e.g. `('module', 'submodule', 'Class', 'function')`, and override the `_table_of_contents_name(node)` method to format the displayed name.

A

**Comment by tony** [comment]
QA of [90c473f4](https://github.com/sphinx-doc/sphinx/pull/10807/commits/90c473f4c4ba09dcb240be1d8740a1233b5bd7be) (Remove `traverse_in_section()`, use `node.findall()`)


<details>

<summary>Screenshot of latest</summary>

image


</details>

<details>

<summary>Either I don't understand the config options, I have a package issue, or they didn't work</summary>

<details>

<summary>add_function_parentheses = False</summary>

_FYI I see no difference with this set_

image

</details>

<details>

<summary>toc_object_entries_parents = 'immediate'</summary>

image

</details>

<details>

<summary>toc_object_entries_parents = 'all'</summary>

image

</details>

</details>

**Comment by AA-Turner** [comment]
@tony please may you run `sphinx-build --version`? I suspect you may not be running the latest changeset as the sidebar shows entries in monospace, which I removed and have not yet re-instated. It should report `sphinx-build 5.2.0+/27bbd09d2`, running from commit [`27bbd09` (#10807)](https://github.com/sphinx-doc/sphinx/pull/10807/commits/27bbd09d2e6507deb770996bf6c0cf9ff10aace4).

An excerpt of your source file would also be useful if at all possible. 

A

**Comment by tony** [comment]
I see sphinx-build 5.2.0+/6d0d695. Maybe I have a bad copy.

<details>

```python
"""Toctree collector for sphinx.environment."""

from typing import Any, Dict, List, Optional, Sequence, Set, Tuple, TypeVar, Union, cast

from docutils import nodes
from docutils.nodes import Element, Node

from sphinx import addnodes
from sphinx.application import Sphinx
from sphinx.environment import BuildEnvironment
from sphinx.environment.adapters.toctree import TocTree
from sphinx.environment.collectors import EnvironmentCollector
from sphinx.locale import __
from sphinx.transforms import SphinxContentsFilter
from sphinx.util import logging, url_re

N = TypeVar('N')

logger = logging.getLogger(__name__)


class TocTreeCollector(EnvironmentCollector):
    def clear_doc(self, app: Sphinx, env: BuildEnvironment, docname: str) -> None:
        env.tocs.pop(docname, None)
        env.toc_secnumbers.pop(docname, None)
        env.toc_fignumbers.pop(docname, None)
        env.toc_num_entries.pop(docname, None)
        env.toctree_includes.pop(docname, None)
        env.glob_toctrees.discard(docname)
        env.numbered_toctrees.discard(docname)

        for subfn, fnset in list(env.files_to_rebuild.items()):
            fnset.discard(docname)
            if not fnset:
                del env.files_to_rebuild[subfn]

    def merge_other(self, app: Sphinx, env: BuildEnvironment, docnames: Set[str],
                    other: BuildEnvironment) -> None:
        for docname in docnames:
            env.tocs[docname] = other.tocs[docname]
            env.toc_num_entries[docname] = other.toc_num_entries[docname]
            if docname in other.toctree_includes:
                env.toctree_includes[docname] = other.toctree_includes[docname]
            if docname in other.glob_toctrees:
                env.glob_toctrees.add(docname)
            if docname in other.numbered_toctrees:
                env.numbered_toctrees.add(docname)

        for subfn, fnset in other.files_to_rebuild.items():
            env.files_to_rebuild.setdefault(subfn, set()).update(fnset & set(docnames))

    def process_doc(self, app: Sphinx, doctree: nodes.document) -> None:
        """Build a TOC from the doctree and store it in the inventory."""
        docname = app.env.docname
        numentries = [0]  # nonlocal again...

        def build_toc(
            node: Union[Element, Sequence[Element]],
            depth: int = 1
        ) -> Optional[nodes.bullet_list]:
            entries: List[Element] = []
            memo_class = '\4\2'  # sentinel, value unimportant
            for sectionnode in node:
                # find all toctree nodes in this section and add them
                # to the toc (just copying the toctree node which is then
                # resolved in self.get_and_resolve_doctree)
                if isinstance(sectionnode, nodes.section):
                    title = sectionnode[0]
                    # copy the contents of the section title, but without references
                    # and unnecessary stuff
                    visitor = SphinxContentsFilter(doctree)
                    title.walkabout(visitor)
                    nodetext = visitor.get_entry_text()
                    anchorname = _make_anchor_name(sectionnode['ids'], numentries)
                    # make these nodes:
                    # list_item -> compact_paragraph -> reference
                    reference = nodes.reference(
                        '', '', internal=True, refuri=docname,
                        anchorname=anchorname, *nodetext)
                    para = addnodes.compact_paragraph('', '', reference)
                    item: Element = nodes.list_item('', para)
                    sub_item = build_toc(sectionnode, depth + 1)
                    if sub_item:
                        item += sub_item
                    entries.append(item)
                # Wrap items under an ``.. only::`` directive in a node for
                # post-processing
                elif isinstance(sectionnode, addnodes.only):
                    onlynode = addnodes.only(expr=sectionnode['expr'])
                    blist = build_toc(sectionnode, depth)
                    if blist:
                        onlynode += blist.children
                        entries.append(onlynode)
                # check within the section for other node types
                elif isinstance(sectionnode, nodes.Element):
                    toctreenode: nodes.Node
                    for toctreenode in sectionnode.findall():
                        if isinstance(toctreenode, nodes.section):
                            continue
                        if isinstance(toctreenode, addnodes.toctree):
                            item = toctreenode.copy()
                            entries.append(item)
                            # important: do the inventory stuff
                            TocTree(app.env).note(docname, toctreenode)
                        # add object signatures within a section to the ToC
                        elif isinstance(toctreenode, addnodes.desc):
                            title = toctreenode.children[0]

                            # Skip entries with no ID (e.g. with ``:noindex:``)
                            if not title['ids']:
                                continue
                            # Skip if no name set. Known not to set 'fullname'
                            # correctly: C/C++/RST domain, option/cmdoption, envvar
                            full_name = title.get('fullname', '')
                            if not full_name:
                                continue

                            # Nest within Python classes
                            if 'class' in title and title['class'] == memo_class:
                                nested_toc = build_toc([toctreenode], depth + 1)
                                if nested_toc:
                                    last_entry: nodes.Element = entries[-1]
                                    if not isinstance(last_entry[-1], nodes.bullet_list):
                                        last_entry.append(nested_toc)
                                    else:
                                        last_entry[-1].extend(nested_toc)
                                continue
                            else:
                                memo_class = full_name

                            # Cut class name from Python methods
                            if toctreenode.get('domain', '') == 'py' and title.get('class'):
                                if full_name.startswith(title['class'] + '.'):
                                    full_name = full_name[len(title['class'] + '.'):]
                            if toctreenode['objtype'] in {'function', 'method'}:
                                full_name += '()'
                            anchorname = _make_anchor_name(title['ids'], numentries)
                            reference = nodes.reference(
                                '', '', nodes.literal('', full_name),
                                internal=True, refuri=docname, anchorname=anchorname)
                            para = addnodes.compact_paragraph('', '', reference)
                            entries.append(nodes.list_item('', para))

            if entries:
                return nodes.bullet_list('', *entries)
            return None

        toc = build_toc(doctree)
        if toc:
            app.env.tocs[docname] = toc
        else:
            app.env.tocs[docname] = nodes.bullet_list('')
        app.env.toc_num_entries[docname] = numentries[0]

    def get_updated_docs(self, app: Sphinx, env: BuildEnvironment) -> List[str]:
        return self.assign_section_numbers(env) + self.assign_figure_numbers(env)

    def assign_section_numbers(self, env: BuildEnvironment) -> List[str]:
        """Assign a section number to each heading under a numbered toctree."""
        # a list of all docnames whose section numbers changed
        rewrite_needed = []

        assigned: Set[str] = set()
        old_secnumbers = env.toc_secnumbers
        env.toc_secnumbers = {}

        def _walk_toc(
            node: Element, secnums: Dict, depth: int, titlenode: Optional[nodes.title] = None
        ) -> None:
            # titlenode is the title of the document, it will get assigned a
            # secnumber too, so that it shows up in next/prev/parent rellinks
            for subnode in node.children:
                if isinstance(subnode, nodes.bullet_list):
                    numstack.append(0)
                    _walk_toc(subnode, secnums, depth - 1, titlenode)
                    numstack.pop()
                    titlenode = None
                elif isinstance(subnode, nodes.list_item):
                    _walk_toc(subnode, secnums, depth, titlenode)
                    titlenode = None
                elif isinstance(subnode, addnodes.only):
                    # at this stage we don't know yet which sections are going
                    # to be included; just include all of them, even if it leads
                    # to gaps in the numbering
                    _walk_toc(subnode, secnums, depth, titlenode)
                    titlenode = None
                elif isinstance(subnode, addnodes.compact_paragraph):
                    numstack[-1] += 1
                    reference = cast(nodes.reference, subnode[0])
                    if depth > 0:
                        number = list(numstack)
                        secnums[reference['anchorname']] = tuple(numstack)
                    else:
                        number = None
                        secnums[reference['anchorname']] = None
                    reference['secnumber'] = number
                    if titlenode:
                        titlenode['secnumber'] = number
                        titlenode = None
                elif isinstance(subnode, addnodes.toctree):
                    _walk_toctree(subnode, depth)

        def _walk_toctree(toctreenode: addnodes.toctree, depth: int) -> None:
            if depth == 0:
                return
            for (_title, ref) in toctreenode['entries']:
                if url_re.match(ref) or ref == 'self':
                    # don't mess with those
                    continue
                elif ref in assigned:
                    logger.warning(__('%s is already assigned section numbers '
                                      '(nested numbered toctree?)'), ref,
                                   location=toctreenode, type='toc', subtype='secnum')
                elif ref in env.tocs:
                    secnums: Dict[str, Tuple[int, ...]] = {}
                    env.toc_secnumbers[ref] = secnums
                    assigned.add(ref)
                    _walk_toc(env.tocs[ref], secnums, depth, env.titles.get(ref))
                    if secnums != old_secnumbers.get(ref):
                        rewrite_needed.append(ref)

        for docname in env.numbered_toctrees:
            assigned.add(docname)
            doctree = env.get_doctree(docname)
            for toctreenode in doctree.findall(addnodes.toctree):
                depth = toctreenode.get('numbered', 0)
                if depth:
                    # every numbered toctree gets new numbering
                    numstack = [0]
                    _walk_toctree(toctreenode, depth)

        return rewrite_needed

    def assign_figure_numbers(self, env: BuildEnvironment) -> List[str]:
        """Assign a figure number to each figure under a numbered toctree."""

        rewrite_needed = []

        assigned: Set[str] = set()
        old_fignumbers = env.toc_fignumbers
        env.toc_fignumbers = {}
        fignum_counter: Dict[str, Dict[Tuple[int, ...], int]] = {}

        def get_figtype(node: Node) -> Optional[str]:
            for domain in env.domains.values():
                figtype = domain.get_enumerable_node_type(node)
                if domain.name == 'std' and not domain.get_numfig_title(node):  # type: ignore
                    # Skip if uncaptioned node
                    continue

                if figtype:
                    return figtype

            return None

        def get_section_number(docname: str, section: nodes.section) -> Tuple[int, ...]:
            anchorname = '#' + section['ids'][0]
            secnumbers = env.toc_secnumbers.get(docname, {})
            if anchorname in secnumbers:
                secnum = secnumbers.get(anchorname)
            else:
                secnum = secnumbers.get('')

            return secnum or ()

        def get_next_fignumber(figtype: str, secnum: Tuple[int, ...]) -> Tuple[int, ...]:
            counter = fignum_counter.setdefault(figtype, {})

            secnum = secnum[:env.config.numfig_secnum_depth]
            counter[secnum] = counter.get(secnum, 0) + 1
            return secnum + (counter[secnum],)

        def register_fignumber(docname: str, secnum: Tuple[int, ...],
                               figtype: str, fignode: Element) -> None:
            env.toc_fignumbers.setdefault(docname, {})
            fignumbers = env.toc_fignumbers[docname].setdefault(figtype, {})
            figure_id = fignode['ids'][0]

            fignumbers[figure_id] = get_next_fignumber(figtype, secnum)

        def _walk_doctree(docname: str, doctree: Element, secnum: Tuple[int, ...]) -> None:
            for subnode in doctree.children:
                if isinstance(subnode, nodes.section):
                    next_secnum = get_section_number(docname, subnode)
                    if next_secnum:
                        _walk_doctree(docname, subnode, next_secnum)
                    else:
                        _walk_doctree(docname, subnode, secnum)
                elif isinstance(subnode, addnodes.toctree):
                    for _title, subdocname in subnode['entries']:
                        if url_re.match(subdocname) or subdocname == 'self':
                            # don't mess with those
                            continue

                        _walk_doc(subdocname, secnum)
                elif isinstance(subnode, nodes.Element):
                    figtype = get_figtype(subnode)
                    if figtype and subnode['ids']:
                        register_fignumber(docname, secnum, figtype, subnode)

                    _walk_doctree(docname, subnode, secnum)

        def _walk_doc(docname: str, secnum: Tuple[int, ...]) -> None:
            if docname not in assigned:
                assigned.add(docname)
                doctree = env.get_doctree(docname)
                _walk_doctree(docname, doctree, secnum)

        if env.config.numfig:
            _walk_doc(env.config.root_doc, ())
            for docname, fignums in env.toc_fignumbers.items():
                if fignums != old_fignumbers.get(docname):
                    rewrite_needed.append(docname)

        return rewrite_needed


def _make_anchor_name(ids: List[str], num_entries: List[int]) -> str:
    if not num_entries[0]:
        # for the very first toc entry, don't add an anchor
        # as it is the file's title anyway
        anchorname = ''
    else:
        anchorname = '#' + ids[0]
    num_entries[0] += 1
    return anchorname


def setup(app: Sphinx) -> Dict[str, Any]:
    app.add_env_collector(TocTreeCollector)

    return {
        'version': 'builtin',
        'parallel_read_safe': True,
        'parallel_write_safe': True,
    }
```

</details>

**Comment by AA-Turner** [comment]
Yep, that isn't the latest -- the 'new' bit should start with `if not toctreenode.get('_toc_name', ''):`

https://github.com/sphinx-doc/sphinx/blob/8d877f1cb7f3dd65cb20a82559ca868e84b12518/sphinx/environment/collectors/toctree.py#L105-L134

A

**Comment by tony** [comment]
@AA-Turner I will fix it and rerun my tests here

**Comment by AA-Turner** [comment]
Thanks!

A

**Comment by tony** [comment]
<details>

<summary>Package Fixed</summary>

Poetry 1.1.15 was at fault - nothing would wipe the old ref, including wiping cache and the virtualenv.

 Manual `pip(1)` worked

```console
pip install -e 'git+https://github.com/AA-Turner/sphinx.git@auto-toc#egg=sphinx'
```

<details>

```
Obtaining sphinx from git+https://github.com/AA-Turner/sphinx.git@auto-toc#egg=sphinx
  WARNING: git clone in ./.venv/src/sphinx exists with URL https://github.com/AA-Turner/sphinx
  WARNING: The plan is to install the git repository https://github.com/AA-Turner/sphinx.git
What to do?  (s)witch, (i)gnore, (w)ipe, (b)ackup w
  WARNING: Deleting ./.venv/src/sphinx
```

</details>

<details>

```
Installing collected packages: sphinx
  Attempting uninstall: sphinx
    Found existing installation: Sphinx 5.2.0.dev20220908
    Uninstalling Sphinx-5.2.0.dev20220908:
      Successfully uninstalled Sphinx-5.2.0.dev20220908
  Running setup.py develop for sphinx
Successfully installed sphinx-5.2.0.dev20220908
```

</details>

</details>


```
libvcs on  sphinx-auto-doc ❯ sphinx-build --version
sphinx-build 5.2.0+/27bbd09d2
```

QA: (#10807) [27bbd09](https://github.com/sphinx-doc/sphinx/pull/10807/commits/27bbd09d2e6507deb770996bf6c0cf9ff10aace4).

<details>

<summary>Defaults</summary>

image

image

</details>


<details>

<summary>add_function_parentheses = False</summary>

image

image

</details>

<details>

<summary>toc_object_entries_parents = 'immediate'</summary>

image

image

</details>

<details>

<summary>toc_object_entries_parents = 'all'</summary>

image

😬 

image

</details>

**Comment by tony** [comment]
@AA-Turner Feel free to rebase and if you would like another QA please let me know

File: tests/test_environment_toctree.py [hunk 2]
```diff
@@ -126,6 +126,46 @@ def test_glob(app):
...
+    assert app.env.toc_num_entries['domains'] == 9
+    assert app.env.toctree_includes['index'] == includefiles
+    for file in includefiles:
+        assert 'index' in app.env.files_to_rebuild[file]
+    assert app.env.glob_toctrees == set()
+    assert app.env.numbered_toctrees == {'index'}
+
+    # tocs
+    toctree = app.env.tocs['domains']
+    print(toctree[0][1][1][1][0])
```
Comment (tony) [inline tests/test_environment_toctree.py]: This and line 164's `print`'s can be cleaned up

**Comment by AA-Turner** [comment]
> This seems worthwhile to document, as a part of the domains documentation somewhere?

Added as [here](https://sphinx--10807.org.readthedocs.build/en/10807/extdev/domainapi.html#sphinx.directives.ObjectDescription._object_hierarchy_parts) and [here](https://sphinx--10807.org.readthedocs.build/en/10807/extdev/domainapi.html#sphinx.directives.ObjectDescription._toc_entry_name).

A

**Comment by AA-Turner** [comment]
> Feel free to rebase and if you would like another QA please let me know

I realised there was a major failing of the previous implementation, multiple signatures defined in the same directive wouldn't be individually listed, for example:

```rest
.. rst:role:: c:member
              c:data
              c:var
              c:func
              <snip>
```

This now [renders correctly (preview docs)](https://sphinx--10807.org.readthedocs.build/en/10807/usage/restructuredtext/domains.html#role-c-member). I also implemented support for the reST domain, especially useful for Sphinx's documentation. 

Another QA would be useful, please!

A

File: sphinx/domains/python.py (hunk 1 above)
Comment (asmeurer) [reply sphinx/domains/python.py:1020]: Was this addressed (documenting the change to `py:module`)?

Comment (AA-Turner) [reply sphinx/domains/python.py:1020]: Yes:

image

A

**Comment by tony** [comment]
@AA-Turner QA On the way

**Comment by tony** [comment]
**QA: #10807 @ [12bd9ec](https://github.com/sphinx-doc/sphinx/pull/10807/commits/12bd9ec66a7be2e78ac3aa96c83d521b46f95a7f)**


```console
pip install -e 'git+https://github.com/AA-Turner/sphinx.git@auto-toc#egg=sphinx'
```

```
sphinx-build --version
sphinx-build 5.2.0+/126bbef
```

<details>

<summary>Defaults</summary>

image

image

</details>


<details>

<summary>add_function_parentheses = False</summary>

image

image


</details>

<details>

<summary>toc_object_entries_show_parents = 'domain'</summary>

image
image

</details>

<details>

<summary>toc_object_entries_show_parents = 'hide'</summary>

image

image


</details>

<details>

<summary>toc_object_entries_show_parents = 'all'</summary>

image

😲

image

</details>

_See also: https://github.com/sphinx-doc/sphinx/pull/10807#issuecomment-1241427754 earlier screenshots from [27bbd09](https://github.com/sphinx-doc/sphinx/pull/10807/commits/27bbd09d2e6507deb770996bf6c0cf9ff10aace4), Sept 8th (4 days earlier)_

**Comment by AA-Turner** [comment]
Thank you Tony. I intend to merge this tomorrow, so if there are any late concerns please raise them sharpish!

A

**Comment by tony** [comment]
@AA-Turner We're on target

P.S. Assuming it's merged, how long until we'll see it published to PyPI?  (looks like this replaces workarounds on a lot of our projects)

**Comment by pradyunsg** [comment]
Hmm... I see stray `.` at the end of the literal name (and before the parens) in the toc, for `toc_object_entries_show_parents = "all"` case.

**Comment by AA-Turner** [comment]
Fixed, hopefully.

A

**Comment by AA-Turner** [comment]
> P.S. Assuming it's merged, how long until we'll see it published to PyPI? (looks like this replaces workarounds on a lot of our projects)

5.2 should be released within the month. I want to clear all items from [the milestone](https://github.com/sphinx-doc/sphinx/milestone/130) before release.

A

**Comment by tony** [comment]
@AA-Turner @pradyunsg Let's try this

[fc82407](https://github.com/sphinx-doc/sphinx/pull/10807/commits/fc82407bc82a39f03e07cb49807e4771606813f0)

<details>

<summary>toc_object_entries_show_parents = 'domain'</summary>

image

image

</details>

<details>

<summary>toc_object_entries_show_parents = 'hide'</summary>

image

image

</details>


<details>

<summary>toc_object_entries_show_parents = 'all'</summary>

image

image


</details>

_See also: [12bd9ec](https://github.com/sphinx-doc/sphinx/pull/10807/commits/12bd9ec66a7be2e78ac3aa96c83d521b46f95a7f) from earlier today_

**Comment by AA-Turner** [comment]
Thanks all for reviews.

A

**Comment by mgeier** [comment]
Thanks for implementing this, this is a great feature!

The structure of the domain object sections (using `<dl>`) is a bit different to "normal" sections (using `<section>`), but with a little tweak (https://github.com/mgeier/insipid-sphinx-theme/pull/121), this works great with my theme (named `insipid`), including emphasizing the currently visible sections in the sidebar TOC.

For a live example, see https://insipid-sphinx-theme.readthedocs.io/en/sphinx5.x/showcase/api-doc.html

**Comment by mgeier** [comment]
Interestingly, the domain object sections get a section number in the TOC (when using the `:numbered:` flag), while the sections in the main text don't get one.

Since the sections in the main text didn't have numbers in the first place, I guess I would expect the TOC entries to have no numbers either?

And the nesting in the reST section isn't correct, but the PR description already mentions that the reST domain is not supported.

**Comment by asmeurer** [comment]
@AA-Turner thank you for doing the work to get this implemented in Sphinx proper. This is going to be a huge benefit to a lot of projects.

**Comment by AA-Turner** [comment]
Thanks for testing this Matthias.

> Interestingly, the domain object sections get a section number in the TOC (when using the `:numbered:` flag), while the sections in the main text don't get one.
> 
> Since the sections in the main text didn't have numbers in the first place, I guess I would expect the TOC entries to have no numbers either?

Please may you open a new issue for this, I'm not sure I fully understand? I'd like to get this fixed before 5.2 release

> And the nesting in the reST section isn't correct, but the PR description already mentions that the reST domain is not supported.

I implemented support for the reST domain so nesting should work -- again please may you open a new issue with a mininal reproducer?

Thanks,
Adam

**Comment by mgeier** [comment]
> Please may you open a new issue for this, I'm not sure I fully understand?

#10826

**Comment by mgeier** [comment]
> I implemented support for the reST domain so nesting should work -- again please may you open a new issue with a mininal reproducer?

#10827

**Comment by domdfcoding** [comment]
It is deliberate that the `target` node and HTML anchor are now *after* the docstring for modules? In the past the `target` was before the docstring, which (for docutils reasons I don't fully understand) made the anchor part of the section heading.

If I apply this patch, the `target` node and the anchor are in the positions they were before:

```patch
diff --git a/sphinx/domains/python.py b/sphinx/domains/python.py
index 6076eb7fb..bd507a21c 100644
--- a/sphinx/domains/python.py
+++ b/sphinx/domains/python.py
@@ -1024,7 +1024,7 @@ class PyModule(SphinxDirective):
             content_node.document = self.state.document
             nested_parse_with_titles(self.state, self.content, content_node)
 
-        ret: List[Node] = [*content_node.children]
+        ret: List[Node] = []
         if not noindex:
             # note module to the domain
             node_id = make_id(self.env, self.state.document, 'module', modname)
@@ -1045,6 +1045,7 @@ class PyModule(SphinxDirective):
             indextext = '%s; %s' % (pairindextypes['module'], modname)
             inode = addnodes.index(entries=[('pair', indextext, node_id, '', None)])
             ret.append(inode)
+        ret.extend(content_node.children)
         return ret
 
     def make_old_id(self, name: str) -> str:
```

**Comment by AA-Turner** [comment]
@domdfcoding please may you open a new issue and tag me? This likely isn't intentional.

A

**Comment by jakobandersen** [comment]
Why was the ``:noindex:`` option added back to the C and C++ domains? They don't support that feature (see https://github.com/sphinx-doc/sphinx/pull/7906 and the originally raised issue).

**Comment by AA-Turner** [comment]
Thanks Jakob--at one point in the PR I was using `noindex` and needed every domain to support it, I later switched to checking for `noindexentry` but clearly forgot to clean this up. The revert for this should go in to 5.2.3, if you don't get to it first I will make a PR over the weekend.

A

**Comment by jakobandersen** [comment]
I don't think ``noindexentry`` is appropriate either, as it's already defined to control whether or not the object goes into the index page, which I believe is orthogonal to this feature. Rather make a new ``notocentry`` option for this control (and a global option for no toc entry on everything, for backwards compatibility).

**Comment by asmeurer** [comment]
IIRC the implementation required handling noindex specially, but maybe that's a detail that can be worked around.

**Pull request merged**